### PR TITLE
[T-163] Simplify diagnostics for plugins

### PR DESCRIPTION
**Pull request by DeMoorJasper**
# ↪️ Pull Request

Rethrow errors from plugins with the correct origin, also pass a logger instance to each plugin call with the correct origin.

Close T-163
Close T-182

## Plugin Loggers

- [x] Resolvers
- [x] Transformers
- [x] Bundlers
- [x] Namers
- [x] Runtimes
- [x] Packagers
- [x] Optimizers
- [x] Reporters
- [x] Validators

## Rethrowing errors

- [x] Resolvers
- [x] Transformers
- [x] Bundlers
- [x] Namers
- [x] Runtimes
- [x] Packagers
- [x] Optimizers
- [x] Reporters
- [x] Validators

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/kitchen-sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/react-hn ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 15.39s | +18.00ms |
| Cached | 10.51s | +145.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 34.53kb | +0.00b | 14.00ms | +1.00ms ⚠️ |

</p></details>

<details><summary>packages/benchmarks/ak-editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 4.09m | +2.83s |
| Cached | 11.36s | +178.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 3.65mb | +115.00b ⚠️ | 2.50m | +2.36s |
| dist/media-viewer.00bf8d7b.js | 86.21kb | +0.00b | 9.57s | +729.00ms ⚠️ |
| dist/card.3e4e47db.js | 75.91kb | +0.00b | 9.57s | +750.00ms ⚠️ |
| dist/popup.e57e7ab5.js | 43.88kb | +0.00b | 49.16s | +4.09s ⚠️ |
| dist/card.f99e1597.js | 5.89kb | +0.00b | 20.21s | +2.03s ⚠️ |
| dist/ResourcedEmojiComponent.2d80adbd.js | 2.25kb | +0.00b | 9.49s | +741.00ms ⚠️ |
| dist/card.9520f336.js | 1.80kb | +0.00b | 20.21s | +2.03s ⚠️ |
| dist/media-viewer-analytics-error-boundary.dd752a9e.js | 777.00b | +0.00b | 20.21s | +2.02s ⚠️ |
| dist/esm.976debcc.js | 178.00b | +0.00b | 9.41s | +712.00ms ⚠️ |
| dist/dropzone.976debcc.js | 177.00b | +1.00b ⚠️ | 2.50m | +1.70m ⚠️ |
| dist/clipboard.976debcc.js | 177.00b | +1.00b ⚠️ | 49.67s | +523.00ms |
| dist/browser.976debcc.js | 177.00b | +1.00b ⚠️ | 1.22m | +23.76s ⚠️ |
| dist/editorView.976debcc.js | 177.00b | -1.00b 🚀 | 49.13s | -22.21s 🚀 |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | -5.00b 🚀 | 51.04s | -1.60m 🚀 |
| dist/16.976debcc.js | 174.00b | -5.00b 🚀 | 49.67s | -340.00ms |
| dist/card.aae6ebb9.js | 62.00b | +0.00b | 20.21s | +2.02s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 3.78mb | +115.00b ⚠️ | 467.00ms | -5.00ms |
| dist/smartMediaEditor.c5927245.js | 597.08kb | +0.00b | 177.00ms | -11.00ms 🚀 |
| dist/EmojiPickerComponent.96e110c7.js | 111.61kb | +0.00b | 136.00ms | -10.00ms 🚀 |
| dist/16.f54dac28.js | 1.34kb | +0.00b | 149.00ms | -8.00ms 🚀 |
| dist/16.74510c56.js | 1.09kb | +0.00b | 149.00ms | -8.00ms 🚀 |
| dist/esm.976debcc.js | 176.00b | +2.00b ⚠️ | 165.00ms | -2.00ms |
| dist/media-card-analytics-error-boundary.976debcc.js | 176.00b | +2.00b ⚠️ | 162.00ms | -3.00ms |
| dist/dropzone.976debcc.js | 176.00b | +2.00b ⚠️ | 158.00ms | -4.00ms |
| dist/clipboard.976debcc.js | 176.00b | +2.00b ⚠️ | 157.00ms | -3.00ms |
| dist/browser.976debcc.js | 176.00b | +2.00b ⚠️ | 157.00ms | -3.00ms |
| dist/16.976debcc.js | 176.00b | +2.00b ⚠️ | 140.00ms | -4.00ms |
| dist/editorView.976debcc.js | 176.00b | +2.00b ⚠️ | 126.00ms | -5.00ms |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/248649155832971785)**

File: packages/core/core/src/ParcelConfig.js [hunk 1]
```diff
@@ -134,8 +134,21 @@ export default class ParcelConfig {
...
+  async getValidators(
+    filePath: FilePath
+  ): Promise<
+    Array<{|
+      name: string,
+      plugin: Validator
+    |}>
+  > {
+    let names = this.getValidatorNames(filePath);
+    return (await this.loadPlugins(names)).map((validator, i) => {
```
Comment (DeMoorJasper) [inline packages/core/core/src/ParcelConfig.js]: Maybe we should just refactor `this.loadPlugins` to return `{name: string, plugin: any}`?

File: packages/core/core/src/ReporterRunner.js [hunk 2]
```diff
@@ -49,7 +49,19 @@ export default class ReporterRunner {
     let reporters = await this.config.getReporters();
 
     for (let reporter of reporters) {
-      await reporter.report(event, this.pluginOptions);
+      try {
+        await reporter.plugin.report({
+          event,
+          options: this.pluginOptions,
+          // Should this even have a logger? Might end up in an infinite loop...
+          logger: new PluginLogger({origin: reporter.name})
```
Comment (devongovett) [inline packages/core/core/src/ReporterRunner.js]: Probably - some reporters won’t be loggers, eg the dev server, bundle visualizers, etc. They might want to log stuff to different reporters.

**Comment by DeMoorJasper** [comment]
@devongovett maybe, @parcel/logger is still being used by the workers

**Pull request merged**